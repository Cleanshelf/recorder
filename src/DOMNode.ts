/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export enum DOMNodeType {
  ELEMENT_NODE = 1,
  ATTRIBUTE_NODE = 2,
  TEXT_NODE = 3,
  CDATA_SECTION_NODE = 4,
  ENTITY_REFERENCE_NODE = 5,
  ENTITY_NODE = 6,
  PROCESSING_INSTRUCTION_NODE = 7,
  COMMENT_NODE = 8,
  DOCUMENT_NODE = 9,
  DOCUMENT_TYPE_NODE = 10,
  DOCUMENT_FRAGMENT_NODE = 11,
  NOTATION_NODE = 12,
}

export default class DOMNode {
  public readonly parentNode: DOMNode | null;
  public readonly nodeName: string;
  public readonly children: DOMNode[];
  public readonly attributes: Map<string, string>;
  public readonly nodeType: DOMNodeType;

  constructor(nodeType, nodeName, attributes = null) {
    this.parentNode = null;
    this.nodeName = nodeName;
    this.children = [];
    this.attributes = new Map();
    this.nodeType = nodeType;

    if (attributes) {
      for (let i = 0; i < attributes.length; i += 2) {
        this.attributes.set(attributes[i], attributes[i + 1]);
      }
    }
  }

  static create(payload) {
    return new DOMNode(payload.nodeType, payload.nodeName, payload.attributes);
  }

  appendChild(node) {
    node.parentNode = this;
    this.children.push(node);
  }

  getAttribute(name) {
    return this.attributes.get(name) || null;
  }

  get classList() {
    const classNames = this.getAttribute("class");
    if (!classNames) return [];
    return classNames.split(/\s+/g);
  }

  get firstChild() {
    return this.children.length ? this.children[0] : null;
  }

  get lastChild() {
    return this.children.length
      ? this.children[this.children.length - 1]
      : null;
  }

  get nextSibling() {
    if (!this.parentNode) return null;

    const index = this.parentNode.children.indexOf(this);
    return this.parentNode.children[index + 1] || null;
  }

  get prevSibling() {
    if (!this.parentNode) return null;

    const index = this.parentNode.children.indexOf(this);
    return index >= 1 ? this.parentNode.children[index - 1] : null;
  }
}
