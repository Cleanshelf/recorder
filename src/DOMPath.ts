/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DOMNode, DOMNodeType } from './dom';

class Step {
  public readonly value: string;
  public readonly optimized: boolean;

  constructor(value: string, optimized: boolean) {
    this.value = value;
    this.optimized = optimized;
  }

  toString() {
    return this.value;
  }
}

function idSelector(id: string) {
  return "#" + id;
}

function cssPathStep(node: DOMNode, isTargetNode: boolean) {
  if (node.nodeType !== DOMNodeType.ELEMENT_NODE) {
    return null;
  }

  const id = node.getAttribute("id");
  if (id) {
    return new Step(idSelector(id), true);
  }

  const nodeNameLower = node.nodeName.toLowerCase();
  if (["html", "body", "head"].includes(nodeNameLower)) {
    return new Step(node.nodeName, true);
  }
  const nodeName = node.nodeName;

  const parent = node.parentNode;
  if (!parent || parent.nodeType === DOMNodeType.DOCUMENT_NODE) {
    return new Step(nodeName, true);
  }

  let needsClassNames = false;
  let needsNthChild = false;
  let ownIndex = -1;
  let elementIndex = -1;
  const siblings = parent.children;
  const ownClassNames = new Set(node.classList);
  for (
    let i = 0;
    (ownIndex === -1 || !needsNthChild) && i < siblings.length;
    i++
  ) {
    const sibling = siblings[i];
    if (sibling.nodeType !== DOMNodeType.ELEMENT_NODE) {
      continue;
    }
    elementIndex += 1;
    if (sibling === node) {
      ownIndex = elementIndex;
      continue;
    }
    if (sibling.nodeName !== nodeName) {
      continue;
    }
    
    needsClassNames = true;
    if (!ownClassNames.size) {
      needsNthChild = true;
      continue;
    }

    const siblingClassNames = new Set(sibling.classList);
    for (const siblingClass of siblingClassNames) {
      if (!ownClassNames.has(siblingClass)) {
        continue;
      }
      ownClassNames.delete(siblingClass);
      if (!ownClassNames.size) {
        needsNthChild = true;
        break;
      }
    }
  }

  let result = nodeName;
  if (
    isTargetNode &&
    nodeName.toLowerCase() === "input" &&
    node.getAttribute("type") &&
    !node.getAttribute("id") &&
    !node.getAttribute("class")
  ) {
    result += `[type="${node.getAttribute("type")}"]`;
  }
  if (needsNthChild) {
    result += `:nth-child(${ownIndex + 1})`;
  } else if (needsClassNames) {
    for (const className of ownClassNames) {
      result += "." + className;
    }
  }
  return new Step(result, false);
}

export function cssPath(node) {
  if (node.nodeType !== DOMNodeType.ELEMENT_NODE) {
    return "";
  }

  const steps = [];
  let currentNode = node;
  while (currentNode) {
    const step = cssPathStep(currentNode, currentNode === node);
    if(!step) {
      break;
    }
    
    steps.push(step);
    if (step.optimized) {
      break;
    }

    currentNode = currentNode.parentNode;
  }

  steps.reverse();
  return steps.join(" > ");
}
