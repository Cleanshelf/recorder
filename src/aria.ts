/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { readFileSync } from 'fs';
import * as path from 'path';

 export function loadAndPatchAriaModule() {
   // Load aria implementation from https://github.com/xi/aria-api (MIT Licensed)
  let script = readFileSync(path.join(__dirname, '../lib/inject.js'), { encoding: 'utf-8' })
  script = script.replace(`var childNodes = [];`, `var childNodes = Array.from(node.shadowRoot?.childNodes || []).filter(n => !getOwner(n) && !isHidden(n))`);
  // Todo(https://github.com/puppeteer/recorder/issues/15): Check if this is the right approach
  script = script.replace(`'input[type="text"]:not([list])',`, `'input[type="text"]:not([list])',\n'input[type="password"]:not([list])',`);

  return script;
 }