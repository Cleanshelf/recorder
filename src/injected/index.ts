/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { getSelector, isSubmitButton } from './dom-helpers';

declare global {
  function addLineToPuppeteerScript(line: string): void;
}

window.addEventListener('click', (e) => {
  // Let submit handle this case if the click is on a submit button
  let currentElement = e.target as HTMLElement;
  while (currentElement) {
    if (isSubmitButton(currentElement)) {
      return;
    }
    currentElement = currentElement.parentElement;
  }

  const selector = getSelector(e.target as HTMLElement);
  addLineToPuppeteerScript(`await click('${selector}');`);
}, true);

window.addEventListener('change', (e) => {
  const value = (e.target as HTMLInputElement).value;
  const escapedValue = value.replace(/'/g, '\\\'');
  const selector = getSelector(e.target as HTMLElement);
  addLineToPuppeteerScript(`await type('${selector}', '${escapedValue}');`);
}, true);

window.addEventListener('submit', (e) => {
  const selector = getSelector(e.target as HTMLElement);
  addLineToPuppeteerScript(`await submit('${selector}');`);
}, true);
