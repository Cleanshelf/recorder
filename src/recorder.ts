/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as puppeteer from 'puppeteer';
import { readFileSync } from 'fs';
import * as path from 'path';

export default async (url: string) => {
  const browser = await puppeteer.launch({
    headless: false,
    defaultViewport: null,
  });

  const page = await browser.pages().then(pages => pages[0]);

  let identation = 0;
  const addLineToPuppeteerScript = (line: string) => {
    const data = '  '.repeat(identation) + line;
    console.log(data);
  }

  page.exposeFunction('addLineToPuppeteerScript', addLineToPuppeteerScript);
  page.evaluateOnNewDocument(readFileSync(path.join(__dirname, '../lib/inject.js'), { encoding: 'utf-8' }));

  // Setup puppeteer
  addLineToPuppeteerScript(`const {open, click, type, submit} = require('@pptr/recorder');`)
  addLineToPuppeteerScript(`open('${url}', {}, async () => {`);
  identation += 1;

  // Open the initial page
  await page.goto(url);

  // Finish the puppeteer script
  return new Promise(resolve => {
    browser.once('disconnected', () => {
      identation -= 1;
      addLineToPuppeteerScript(`})`);
      resolve();
    });
  });
}
