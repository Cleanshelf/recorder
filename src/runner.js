/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const puppeteer = require('puppeteer');
const fs = require('fs');
const timers = require('timers');

const aria = fs.readFileSync('./node_modules/aria-api/dist/aria.js', { encoding: 'utf8' });

const ariaName = new Function('element', 'selector', `
  // Inject the aria library in case it has not been loaded yet
  if(!globalThis.aria) {${aria}}

  const elements = document.getElementsByTagName('*');
  for(const element of elements) {
    if(!element.parentElement) continue;
    if(aria.getName(element) === selector) {
      return element;
    }
  }

  return null;
`);

puppeteer.__experimental_registerCustomQueryHandler('ariaName', ariaName);

const ariaNameContains = new Function('element', 'selector', `
  // Inject the aria library in case it has not been loaded yet
  if(!globalThis.aria) {${aria}}

  const elements = document.getElementsByTagName('*');
  for(const element of elements) {
    if(!element.parentElement) continue;
    if(aria.getName(element).includes(selector)) {
      return element;
    }
  }

  return null;
`);

puppeteer.__experimental_registerCustomQueryHandler('ariaNameContains', ariaNameContains);

const timeout = t => new Promise(cb => timers.setTimeout(cb, t));

const debug = async (selector) => {
  console.log(await page.$eval(selector, e => ({
    tagName: e.tagName,
    attributes: Array.from(e.attributes).reduce((prev, curr) => ({ ...prev, [curr.nodeName]: [curr.nodeValue] }), {}),
  })));
}

let browser, page;
const delay = 100;
module.exports = {
  async open(url, cb) {
    browser = await puppeteer.launch({ headless: false, defaultViewport: null });
    const pages = await browser.pages();
    page = pages[0];
    await page.evaluateOnNewDocument(aria);
    await page.goto(url);
    await timeout(1000);
    await cb(page, browser);
    await browser.close();
  },
  async click(selector) {
    await timeout(delay);
    console.log('click', selector);
    const element = await page.waitForSelector(selector);
    await element.click();
  },
  async type(selector, value) {
    await timeout(delay);
    console.log('type', selector, value);
    const element = await page.waitForSelector(selector);
    await element.click({ clickCount: 3 });
    await element.press('Backspace');
    await element.type(value);
  },
  async submit(selector) {
    await timeout(delay);
    console.log('submit', selector);
    await page.$eval(selector, form => form.requestSubmit());
  }
}
