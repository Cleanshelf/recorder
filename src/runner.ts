/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as fs from 'fs';
import * as timers from 'timers';
import * as path from 'path';
import * as puppeteer from 'puppeteer';
import * as readline from 'readline';

declare const __dirname;

const aria = fs.readFileSync(path.join(__dirname, '../node_modules/aria-api/dist/aria.js'), { encoding: 'utf8' });

const ariaName = new Function('element', 'selector', `
  // Inject the aria library in case it has not been loaded yet
  if(!globalThis.aria) {${aria}}

  const elements = document.getElementsByTagName('*');
  for(const element of elements) {
    if(!element.parentElement) continue;
    if(aria.getName(element) === selector) {
      return element;
    }
  }

  return null;
`);

puppeteer.__experimental_registerCustomQueryHandler('ariaName', ariaName);

const ariaNameContains = new Function('element', 'selector', `
  // Inject the aria library in case it has not been loaded yet
  if(!globalThis.aria) {${aria}}

  const elements = document.getElementsByTagName('*');
  for(const element of elements) {
    if(!element.parentElement) continue;
    if(aria.getName(element).includes(selector)) {
      return element;
    }
  }

  return null;
`);

puppeteer.__experimental_registerCustomQueryHandler('ariaNameContains', ariaNameContains);

const timeout = t => new Promise(cb => timers.setTimeout(cb, t));

let browser, page;
let delay = 100;
const debug = process.env.DEBUG;

interface RunnerOptions {
  delay: number
}

async function beforeStep(...args) {
  console.log(...args);

  if(!debug) {
    await timeout(delay);
    return;
  }
  
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  await new Promise(resolve => rl.question('Press enter to execute this step?', ans => {
    rl.close();
    resolve(ans);
  }));
}

export async function open(url, options: RunnerOptions, cb) {
  delay = options.delay || 100;
  browser = await puppeteer.launch({ headless: false, defaultViewport: null });
  const pages = await browser.pages();
  page = pages[0];
  await page.evaluateOnNewDocument(aria);
  await page.goto(url);
  await timeout(1000);
  await cb(page, browser);
  await browser.close();
}

export async function click(selector) {
  await beforeStep('click', selector);
  const element = await page.waitForSelector(selector);
  await element.click();
}

export async function type(selector, value) {
  await beforeStep('type', selector, value);
  const element = await page.waitForSelector(selector);
  await element.click({ clickCount: 3 });
  await element.press('Backspace');
  await element.type(value);
}

export async function submit(selector) {
  await beforeStep('submit', selector);
  await page.$eval(selector, form => form.requestSubmit());
}
