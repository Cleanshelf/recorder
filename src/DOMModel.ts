/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import DOMNode from './DOMNode';

export default class DOMModel {
  public document: DOMNode | null;
  private _nodesById: Map<number, DOMNode>;

  constructor() {
    this._nodesById = new Map();
    this.document = null;
  }

  load(document) {
    this.loadNode(document.root);
  }

  loadNode = (root) => {
    const stack = [root];
    while (stack.length) {
      const data = stack.shift();

      const node = DOMNode.create(data);
      this._nodesById.set(data.nodeId, node);
      if (!this.document) {
        this.document = node;
      }

      if (data.parentId) {
        const parent = this._nodesById.get(data.parentId);
        parent.appendChild(node);
      }

      if (data.children) stack.push(...data.children);
      if (data.shadowRoots) stack.push(...data.shadowRoots);
      if (data.pseudoElements) stack.push(...data.pseudoElements);
    }
  };

  setChildNodes({parentId, nodes}) {
    const parent = this._nodesById.get(parentId);
    if (!parent) return;

    for (const node of nodes) {
      this.loadNode(node);
    }
  }

  get(id) {
    return this._nodesById.get(id);
  }

  toString() {
    let depth = 0;
    const output = [];
    function append(line) {
      output.push("  ".repeat(depth) + line);
    }
    let currentNode = this.document;
    while (currentNode) {
      append(currentNode.nodeName);
      if (currentNode.firstChild) {
        depth += 1;
        currentNode = currentNode.firstChild;
      } else if (currentNode.nextSibling) {
        currentNode = currentNode.nextSibling;
      } else if (currentNode.parentNode) {
        depth -= 1;
        currentNode = currentNode.parentNode.nextSibling;
      } else {
        break;
      }
    }

    return output.join("\n");
  }
}
