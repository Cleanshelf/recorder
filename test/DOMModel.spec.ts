/**
 * Copyright 2020 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import DOMModel from '../src/DOMModel';

import getDocumentPayload from './fixtures/getDocument';
import setChildNodesPayload from './fixtures/setChildNodes';
import { DOMNode } from '../src/dom';

describe('DOMMode', () => {
  it('should create a new DOM model', () => {
    const model = new DOMModel();

    expect(model.document).toBe(null)
  });

  it('should load a new DOM document from payload', () => {
    const model = new DOMModel();
    model.load(getDocumentPayload);

    expect(model.document).toBeInstanceOf(DOMNode);
  });

  it('update child nodes from document payload', () => {
    const model = new DOMModel();
    model.load(getDocumentPayload);
    model.setChildNodes(setChildNodesPayload);

    expect(model.document).toBeInstanceOf(DOMNode);
  });

  it('should return a single node by id', () => {
    const model = new DOMModel();
    model.load(getDocumentPayload);

    const node = model.get(5);

    expect(node).toBeInstanceOf(DOMNode);
    expect(node.nodeName).toBe('BODY');
  });

  it('should return a string representing the dom tree', () => {
    const model = new DOMModel();
    model.load(getDocumentPayload);

    expect(model.toString()).toMatchSnapshot();
  });
});
