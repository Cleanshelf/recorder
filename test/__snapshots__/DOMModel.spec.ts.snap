// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOMMode should return a string representing the dom tree 1`] = `
"#document
  html
  HTML
    HEAD
    BODY"
`;
